### PR TITLE
mgr: add type annotations to BaseMgrModule and influx

### PR DESCRIPTION
**Pull request by tchaikov**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/mgr/ActivePyModules.cc [hunk 1]
```diff
@@ -63,67 +63,6 @@ ActivePyModules::ActivePyModules(
 
 ActivePyModules::~ActivePyModules() = default;
 
-namespace {
```
Comment (sebastian-philipp) [inline src/mgr/ActivePyModules.cc:66]: This seems somewhat unrelated to this PR. you want to split it up into a new PR?

File: src/mgr/BaseMgrModule.cc [hunk 2]
```diff
@@ -490,9 +490,13 @@ ceph_set_module_option(BaseMgrModule *self, PyObject *args)
   if (value) {
     val = value;
   }
-  without_gil([&] {
-    self->py_modules->set_config(module, key, val);
+  auto [ret, msg] = without_gil([&] {
+    return self->py_modules->set_config(module, key, val);
   });
+  if (ret) {
+    PyErr_SetString(PyExc_ValueError, msg.c_str());
```
Comment (sebastian-philipp) [inline src/mgr/BaseMgrModule.cc:497]: you want to add this to the docs? https://docs.ceph.com/en/latest/mgr/modules/?#mgr_module.MgrModule.set_module_option

do you want to add a test in mgr/selftest to verify an exeption is raised?

File: src/pybind/mgr/influx/module.py [hunk 3]
```diff
@@ -387,7 +387,7 @@ def shutdown(self) -> None:
         self.log.debug('Shutting down queue workers')
 
         for _ in self.workers:
-            self.queue.put(None)
+            self.queue.put([])
```
Comment (sebastian-philipp) [inline src/pybind/mgr/influx/module.py:390]: This changes the logic. `None` was used as a special value to indicate worker shut down. The previous value `[]` has now become a special case. 

I'd :+1: to keep the original behavior.

File: src/pybind/mgr/ceph_module.pyi [hunk 4]
```diff
@@ -2,6 +2,8 @@
 # Used by mypy to do proper type checking of mgr modules.
 # Without this file, all classes have undefined base classes.
 
+from typing import Any, Dict, List, Optional, Union
```
Comment (sebastian-philipp) [inline src/pybind/mgr/ceph_module.pyi:5]: Adding argument types to `ceph_module.pyi` only provides moderate gain,  as all of those methods are wrapped in mgr_module already.

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

File: src/mgr/ActivePyModules.cc (hunk 1 above)
Comment (tchaikov) [reply src/mgr/ActivePyModules.cc:66]: i created #39610. but that PR conflicts with this one. i would like to get it merged first.

File: src/pybind/mgr/ceph_module.pyi (hunk 4 above)
Comment (tchaikov) [reply src/pybind/mgr/ceph_module.pyi:5]: yeah, i knew.

**Comment by tchaikov** [comment]
jenkins test api

**Comment by tchaikov** [comment]
jenkins test api

**Comment by tchaikov** [comment]
@sebastian-philipp mind taking another look?

**Review by sebastian-philipp — APPROVED** [review]

**Comment by sebastian-philipp** [comment]
I'm still a bit worried about https://github.com/ceph/ceph/pull/39602#discussion_r580103795 you're sure we're not introducing an edge case here?

File: src/pybind/mgr/influx/module.py (hunk 3 above)
Comment (tchaikov) [reply src/pybind/mgr/influx/module.py:390]: @sebastian-philipp you are right. there is chance that we have no pools in cluster in that case, we should not signal the workers to stop and exit. unless we never push an empty list of data points to the queue if there is no data to push. but i don't want to change the existing behavior which writes empty data points to influx db even if there is no data.

so i changed the type of `self.queue` to `queue.Queue[Optional[List[Dict[str, str]]]]`. not a good example of conciseness, but at least it's correct =)

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-02-20_09:53:59-rados-wip-kefu-testing-2021-02-20-1628-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-02-21_03:27:10-rados-wip-kefu-testing-2021-02-20-1628-distro-basic-smithi/

**Comment by tobias-urdin** [comment]
@tchaikov catching this https://github.com/ceph/ceph/pull/39602/commits/e457ca50011f70cf01a62323998af233a484f338 issue in 15.2.11 but can see it's not backported to Pacific either.

Happens when a OSD is purged at the same time it tries to loop and gather daemon stats.

```
2021-09-30T15:14:17.568+0200 7febc5e56700 -1 mgr get_metadata_python Requested missing service osd.133
2021-09-30T15:14:17.582+0200 7febc5e56700 -1 log_channel(cluster) log [ERR] : Unhandled exception from module 'influx' while running on mgr.osc-cloudstore2-sto1-prod: 'NoneType' object is not subscriptable
2021-09-30T15:14:17.582+0200 7febc5e56700 -1 influx.serve:
2021-09-30T15:14:17.582+0200 7febc5e56700 -1 Traceback (most recent call last):
  File "/usr/share/ceph/mgr/influx/module.py", line 484, in serve
    self.send_to_influx()
  File "/usr/share/ceph/mgr/influx/module.py", line 389, in send_to_influx
    for chunk in self.chunk(points, self.config['batch_size']):
  File "/usr/share/ceph/mgr/influx/module.py", line 120, in chunk
    xs.append(next(l))
  File "/usr/share/ceph/mgr/influx/module.py", line 269, in get_daemon_stats
    "host": metadata['hostname'],
TypeError: 'NoneType' object is not subscriptable
```

Any chance this will be backported to Pacific/Octopus?

**Comment by tchaikov** [comment]
@tobias-urdin hi Tobias, thanks for reporting this issue! i just created #43368 and #43369 to address it.
